Cython/Compiler/ModuleNode.py

```diff
@@ -3711,6 +3791,34 @@ def generate_c_variable_export_code(self, env, code):
                 f'__Pyx_ExportVoidPtr({api_dict}, {name}, (void *)&{entry.cname}, "{signature}")'
             )
 
+    def generate_c_shared_function_export_code(self, code, shared_function_definitions: Sequence[Code.SharedFunctionDecl]):
+        api_dict = code.funcstate.allocate_temp(py_object_type, manage_ref=True)
+        if shared_function_definitions:
+            code.globalstate.use_utility_code(
+                UtilityCode.load_cached("GetApiDict", "ImportExport.c"))
+        code.putln(
+            f"{api_dict} = __Pyx_ApiExport_GetApiDict(); "
+            f"{code.error_goto_if_null(api_dict, self.pos)}"
+        )
+        code.put_gotref(api_dict, py_object_type)
+
+        code.globalstate.use_utility_code(
+            UtilityCode.load_cached("FunctionExport", "ImportExport.c"))
+
+        for shared_func_def in shared_function_definitions:
+            func_name = shared_func_def.name
+            func_params = shared_func_def.params
+            func_ret_type = shared_func_def.ret
+            func_name_from_mstate = code.intern_identifier(EncodedString(func_name))
+            code.put_error_if_neg(
+                self.pos,
+                f'__Pyx_ExportFunction({api_dict}, {func_name_from_mstate}, (void (*)(void)){func_name}, "{func_ret_type}({func_params})")'
+            )
+
+        code.put_decref_clear(api_dict, py_object_type)
+        code.funcstate.release_temp(api_dict)
```

Comment:
This method doesn't look specific to the class that owns it. If you can move it (and the other generation method) over into the `SharedUtilityExporter`, then I think you can avoid passing down the callables.